### ISSUE TITLE
[Unable to Send Message for Newly Subscribed P2P Topic]

### ISSUE THREAD

**Issue by riandyrn** [tinode/example-react-js#4]
Hello,

I found bug in sending message for newly subscribed user to P2P topic. The user won’t be able to send message due to insufficient default permission (missing ‘W’ permission).

You could reproduce this bug by using these steps:
- create new user
- create new chat, find someone (I’m using bob —> email:bob@example.com)
- start chat, then try to send message
- open chrome dev console
- you will find 403 error

The reason why this error happens is because when client send “sub” command to P2P topic without specifying the permission request, the default value for acs.want is “JRPD”, yet server give “JRWP”, so the resulted acs.mode is “JRP” (missing ‘W’ permission). With this mode user doesn’t have enough permission to send message to the peer.

> [06:58:37:169] out: {"sub":{"id":"86972","topic":"usrlDBwwM89EXs","get":{"what":"data sub desc"}}}
> 
> [06:58:37:464] in: {"ctrl":{"id":"86972","topic":"usrlDBwwM89EXs","params":{"acs":{"want":"JRPD","given":"JRWP","mode":"JRP"}},"code":200,"text":"ok","ts":"2017-07-30T06:58:37.296Z"}}

You could see the more details on dev console.

Thanks.

**Comment by or-else** [comment]
I assume this is with stock database, no customisation of any kind?

**Comment by riandyrn** [comment]
Yep, I just tried earlier by using http://api.tinode.co/x/example-react-js/

**Comment by or-else** [comment]
I cannot reproduce: 

```
[07:21:34:962] Connecting to: ws://localhost:6060/v0/channels?apikey=AQEAAAABAAD_rAp4DJh05a1HAwFT3A6K
tinode.js:701 [07:21:34:966] out: {"hi":{"id":"95814","ver":"0.13","ua":"TinodeWeb/0.13 (MacIntel) tinodejs/0.13"}}
tinode.js:701 [07:21:34:967] out: {"acc":{"id":"95815","user":"new","scheme":"basic","secret":"YWRhbTphZGFtMTIz","login":true,"desc":{"public":{"fn":"Adam 123"}}}}
tinode.js:701 [07:21:34:968] in: {"ctrl":{"id":"95814","params":{"ver":"0.13"},"code":201,"text":"created","ts":"2017-07-30T07:21:34.968Z"}}
tinode.js:701 [07:21:35:092] in: {"ctrl":{"id":"95815","params":{"authlvl":"auth","desc":{"created":"2017-07-30T07:21:34.97Z","updated":"2017-07-30T07:21:34.97Z","defacs":{"auth":"JRPD","anon":"N"},"public":{"fn":"Adam 123"}},"expires":"2017-07-30T10:21:35.092371704+03:00","token":"nJas6q8AVLr/iH1ZFAAAAHA6FIwQ3FfY83Xiyau2l//JPXKzRxtqijwtium0Nycr","user":"usrnJas6q8AVLo"},"code":201,"text":"created","ts":"2017-07-30T07:21:34.968Z"}}
tinode.js:701 [07:21:35:102] out: {"sub":{"id":"95816","topic":"me","get":{"what":"data sub desc"}}}
tinode.js:701 [07:21:35:107] in: {"ctrl":{"id":"95816","topic":"me","params":{"acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"}},"code":200,"text":"ok","ts":"2017-07-30T07:21:35.105Z"}}
tinode.js:701 [07:21:35:107] in: {"meta":{"id":"95816","topic":"me","ts":"2017-07-30T07:21:35.105Z","desc":{"created":"2017-07-30T07:21:34.97Z","updated":"2017-07-30T07:21:34.97Z","defacs":{"auth":"JRPD","anon":"N"},"public":{"fn":"Adam 123"}}}}
tinode.js:701 [07:21:35:112] in: {"meta":{"id":"95816","topic":"me","ts":"2017-07-30T07:21:35.106Z"}}
tinode.js:701 [07:21:35:112] in: {"ctrl":{"id":"95816","topic":"me","code":200,"text":"ok","ts":"2017-07-30T07:21:35.107Z"}}
tinode.js:701 [07:21:42:805] out: {"sub":{"id":"95817","topic":"fnd","get":{"what":"sub desc"}}}
tinode.js:701 [07:21:42:814] in: {"ctrl":{"id":"95817","topic":"fnd","params":{"acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"}},"code":200,"text":"ok","ts":"2017-07-30T07:21:42.809Z"}}
tinode.js:701 [07:21:42:814] in: {"meta":{"id":"95817","topic":"fnd","ts":"2017-07-30T07:21:42.809Z","desc":{"created":"2017-07-30T07:21:34.97Z","updated":"2017-07-30T07:21:34.97Z","acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"}}}}
tinode.js:701 [07:21:42:815] in: {"meta":{"id":"95817","topic":"fnd","ts":"2017-07-30T07:21:42.809Z"}}
tinode.js:701 [07:21:56:777] out: {"set":{"id":"95818","topic":"fnd","desc":{"private":["email:bob@example.com"]}}}
tinode.js:701 [07:21:56:790] in: {"ctrl":{"id":"95818","topic":"fnd","code":200,"text":"ok","ts":"2017-07-30T07:21:56.779Z"}}
tinode.js:701 [07:21:56:793] out: {"get":{"id":"95819","topic":"fnd","what":"sub desc"}}
tinode.js:701 [07:21:56:794] in: {"meta":{"id":"95819","topic":"fnd","ts":"2017-07-30T07:21:56.794Z","desc":{"created":"2017-07-30T07:21:34.97Z","updated":"2017-07-30T07:21:34.97Z","acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"},"private":["email:bob@example.com"]}}}
tinode.js:701 [07:21:56:797] in: {"meta":{"id":"95819","topic":"fnd","ts":"2017-07-30T07:21:56.794Z","sub":[{"updated":"2017-07-24T13:16:38.875Z","acs":{"mode":"JRWPS"},"public":{"fn":"Bob Smith","photo":{"data":"<7236, bytes: /9j/4AAQSkZJ...30e375//2Q==>","type":"jpg"}},"private":["email:bob@example.com"],"user":"usravVGHLCBbKo"}]}}
tinode.js:701 [07:21:58:033] out: {"sub":{"id":"95820","topic":"usravVGHLCBbKo","get":{"what":"data sub desc"}}}
tinode.js:701 [07:21:58:074] in: {"ctrl":{"id":"95820","topic":"usravVGHLCBbKo","params":{"acs":{"want":"JRWP","given":"JRWP","mode":"JRWP"}},"code":200,"text":"ok","ts":"2017-07-30T07:21:58.038Z"}}
tinode.js:701 [07:21:58:075] in: {"meta":{"id":"95820","topic":"usravVGHLCBbKo","ts":"2017-07-30T07:21:58.074Z","desc":{"created":"2017-07-30T07:21:58.038Z","updated":"2017-07-30T07:21:58.038Z","acs":{"want":"JRWP","given":"JRWP","mode":"JRWP"},"public":{"fn":"Bob Smith","photo":{"data":"<7236, bytes: /9j/4AAQSkZJ...30e375//2Q==>","type":"jpg"}}}}}
tinode.js:701 [07:21:58:085] in: {"meta":{"id":"95820","topic":"usravVGHLCBbKo","ts":"2017-07-30T07:21:58.074Z","sub":[{"updated":"2017-07-30T07:21:58.038Z","acs":{"mode":"JRWP"},"public":{"fn":"Adam 123"},"user":"usrnJas6q8AVLo"},{"updated":"2017-07-30T07:21:58.038Z","acs":{"mode":"JRWP"},"public":{"fn":"Bob Smith","photo":{"data":"<7236, bytes: /9j/4AAQSkZJ...30e375//2Q==>","type":"jpg"}},"user":"usravVGHLCBbKo"}]}}
Inline Babel script:2013 topic subscriptions updated
tinode.js:701 [07:21:58:088] in: {"ctrl":{"id":"95820","topic":"usravVGHLCBbKo","code":200,"text":"ok","ts":"2017-07-30T07:21:58.080Z"}}
tinode.js:701 [07:21:59:658] out: {"note":{"topic":"usravVGHLCBbKo","what":"kp"}}
tinode.js:701 [07:22:01:218] out: {"pub":{"id":"95821","topic":"usravVGHLCBbKo","content":"hi bob"}}
tinode.js:701 [07:22:01:236] in: {"ctrl":{"id":"95821","topic":"usravVGHLCBbKo","params":{"seq":1},"code":202,"text":"accepted","ts":"2017-07-30T07:22:01.219Z"}}
tinode.js:701 [07:22:01:236] in: {"data":{"topic":"usravVGHLCBbKo","from":"usrnJas6q8AVLo","ts":"2017-07-30T07:22:01.219Z","seq":1,"content":"hi bob"}}
tinode.js:701 [07:22:01:747] out: {"note":{"topic":"usravVGHLCBbKo","what":"recv","seq":1}}
tinode.js:701 [07:22:02:244] out: {"note":{"topic":"usravVGHLCBbKo","what":"read","seq":1}}
```

**Comment by or-else** [comment]
It's probably a misconfigured DB. Try regenerating it first.

**Comment by or-else** [comment]
Or see if you can reproduce the steps which lead to DB misconfiguration.

**Comment by riandyrn** [comment]
Hmm.., that's weird. I'm using the live example app (http://api.tinode.co/x/example-react-js/)

I notice that you point to your localhost instead of ws://api.tinode.co/v0/channels?apikey=AQEAAAABAAD_rAp4DJh05a1HAwFT3A6K, maybe because you haven't push the update?

Here is my full log on the latest try:

> [07:30:47:566] Connecting to: ws://api.tinode.co/v0/channels?apikey=AQEAAAABAAD_rAp4DJh05a1HAwFT3A6K
tinode.js:677 [07:30:48:141] out: {"hi":{"id":"70564","ver":"0.13","ua":"TinodeWeb/0.13 (MacIntel) tinodejs/0.13"}}
tinode.js:677 [07:30:48:142] out: {"acc":{"id":"70565","user":"new","scheme":"basic","secret":"aGVsbG86aGVsbG8xMjM=","login":true,"desc":{"public":{"fn":"hello"}}}}
tinode.js:677 [07:30:48:486] in: {"ctrl":{"id":"70564","params":{"build":"20170711T06:44:52Z","ver":"0.13"},"code":201,"text":"created","ts":"2017-07-30T07:30:48.259Z"}}
tinode.js:677 [07:30:48:524] in: {"ctrl":{"id":"70565","params":{"desc":{"created":"2017-07-30T07:30:48.26Z","updated":"2017-07-30T07:30:48.26Z","defacs":{"auth":"JRPD","anon":"N"},"public":{"fn":"hello"}},"expires":"2017-08-13T07:30:48.259Z","token":"FktuRip73dEoAJBZ5GPn6ZqpBTuAvexIK3cJ0EYhBrmPvIB7af5PICzwiGM=","user":"usrFktuRip73dE"},"code":201,"text":"created","ts":"2017-07-30T07:30:48.259Z"}}
tinode.js:677 [07:30:48:533] out: {"sub":{"id":"70566","topic":"me","get":{"what":"data sub desc"}}}
tinode.js:677 [07:30:48:895] in: {"ctrl":{"id":"70566","topic":"me","params":{"acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"}},"code":200,"text":"ok","ts":"2017-07-30T07:30:48.638Z"}}
tinode.js:677 [07:30:48:896] in: {"meta":{"id":"70566","topic":"me","ts":"2017-07-30T07:30:48.638Z","desc":{"created":"2017-07-30T07:30:48.26Z","updated":"2017-07-30T07:30:48.26Z","defacs":{"auth":"JRPD","anon":"N"},"public":{"fn":"hello"}}}}
tinode.js:677 [07:30:48:899] in: {"meta":{"id":"70566","topic":"me","ts":"2017-07-30T07:30:48.639Z"}}
tinode.js:677 [07:30:48:900] in: {"ctrl":{"id":"70566","topic":"me","code":200,"text":"ok","ts":"2017-07-30T07:30:48.640Z"}}
tinode.js:677 [07:30:54:090] out: {"sub":{"id":"70567","topic":"fnd","get":{"what":"sub desc"}}}
tinode.js:677 [07:30:54:438] in: {"ctrl":{"id":"70567","topic":"fnd","params":{"acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"}},"code":200,"text":"ok","ts":"2017-07-30T07:30:54.213Z"}}
tinode.js:677 [07:30:54:438] in: {"meta":{"id":"70567","topic":"fnd","ts":"2017-07-30T07:30:54.213Z","desc":{"created":"2017-07-30T07:30:48.26Z","updated":"2017-07-30T07:30:48.26Z","acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"}}}}
tinode.js:677 [07:30:54:439] in: {"meta":{"id":"70567","topic":"fnd","ts":"2017-07-30T07:30:54.213Z"}}
tinode.js:677 [07:31:01:873] out: {"set":{"id":"70568","topic":"fnd","desc":{"private":["email:bob@example.com"]}}}
tinode.js:677 [07:31:02:179] in: {"ctrl":{"id":"70568","topic":"fnd","code":200,"text":"ok","ts":"2017-07-30T07:31:01.991Z"}}
tinode.js:677 [07:31:02:186] out: {"get":{"id":"70569","topic":"fnd","what":"sub desc"}}
tinode.js:677 [07:31:02:545] in: {"meta":{"id":"70569","topic":"fnd","ts":"2017-07-30T07:31:02.307Z","desc":{"created":"2017-07-30T07:30:48.26Z","updated":"2017-07-30T07:30:48.26Z","acs":{"want":"JRPD","given":"JRPD","mode":"JRPD"},"private":["email:bob@example.com"]}}}
tinode.js:677 [07:31:02:548] in: {"meta":{"id":"70569","topic":"fnd","ts":"2017-07-30T07:31:02.307Z","sub":[{"updated":"2017-07-24T09:15:03.669Z","acs":{"mode":"JRWPS"},"public":{"fn":"Bob Smith","photo":{"data":"<7236, bytes: /9j/4AAQSkZJ...30e375//2Q==>","type":"jpg"}},"private":["email:bob@example.com"],"user":"usrlDBwwM89EXs"}]}}
tinode.js:677 [07:31:03:915] out: {"sub":{"id":"70570","topic":"usrlDBwwM89EXs","get":{"what":"data sub desc"}}}
tinode.js:677 [07:31:04:222] in: {"ctrl":{"id":"70570","topic":"usrlDBwwM89EXs","params":{"acs":{"want":"JRPD","given":"JRWP","mode":"JRP"}},"code":200,"text":"ok","ts":"2017-07-30T07:31:04.033Z"}}
tinode.js:677 [07:31:04:226] in: {"meta":{"id":"70570","topic":"usrlDBwwM89EXs","ts":"2017-07-30T07:31:04.044Z","desc":{"created":"2017-07-30T07:31:04.033Z","updated":"2017-07-30T07:31:04.033Z","acs":{"want":"JRPD","given":"JRWP","mode":"JRP"},"public":{"fn":"Bob Smith","photo":{"data":"<7236, bytes: /9j/4AAQSkZJ...30e375//2Q==>","type":"jpg"}}}}}
tinode.js:677 [07:31:04:564] in: {"meta":{"id":"70570","topic":"usrlDBwwM89EXs","ts":"2017-07-30T07:31:04.044Z","sub":[{"updated":"2017-07-30T07:31:04.033Z","acs":{"mode":"JRWP"},"public":{"fn":"Bob Smith","photo":{"data":"<7236, bytes: /9j/4AAQSkZJ...30e375//2Q==>","type":"jpg"}},"user":"usrlDBwwM89EXs"},{"updated":"2017-07-30T07:31:04.033Z","acs":{"mode":"JRP"},"public":{"fn":"hello"},"user":"usrFktuRip73dE"}]}}
Inline Babel script:1867 topic subscriptions updated
tinode.js:677 [07:31:04:565] in: {"data":{"topic":"me","from":"usrFktuRip73dE","ts":"2017-07-30T07:31:04.043Z","seq":1,"content":{"acs":{"given":"JRWP","mode":"JRP","want":"JRPD"},"act":"upd","topic":"usrlDBwwM89EXs","user":"usrFktuRip73dE"}}}
Inline Babel script:1067 Got invite:
Inline Babel script:1068 {topic: "me", from: "usrFktuRip73dE", ts: Sun Jul 30 2017 14:31:04 GMT+0700 (WIB), seq: 1, content: {…}}
tinode.js:677 [07:31:04:569] in: {"ctrl":{"id":"70570","topic":"usrlDBwwM89EXs","code":200,"text":"ok","ts":"2017-07-30T07:31:04.048Z"}}
tinode.js:677 [07:31:12:012] out: {"note":{"topic":"usrlDBwwM89EXs","what":"kp"}}
tinode.js:677 [07:31:12:641] out: {"pub":{"id":"70571","topic":"usrlDBwwM89EXs","content":"Test"}}
tinode.js:677 [07:31:12:959] in: {"ctrl":{"id":"70571","topic":"usrlDBwwM89EXs","code":403,"text":"permission denied","ts":"2017-07-30T07:31:12.765Z"}}
tinode.js:735 Uncaught (in promise) Error: 403 permission denied
    at execPromise (tinode.js:735)
    at Object.dispatchMessage [as onMessage] (tinode.js:921)
    at WebSocket.conn.onmessage (tinode.js:406)
execPromise @ tinode.js:735
dispatchMessage @ tinode.js:921
conn.onmessage @ tinode.js:406

**Comment by or-else** [comment]
You are running your tests against api.tinode.co. It's not a good environment to run tests against. It's open to public without restrictions. Most likely someone played with Bob's permissions.

**Comment by or-else** [comment]
Try running your own server. Then you can be sure that issues are reproducible.

**Comment by riandyrn** [comment]
Hmm.., yep, you’re right. After I pull the latest change from devel the bug is gone.

Previously I didn’t aware you already committed many changes (sorry my mistake :P), so my local was still having the same bug. At first I thought that the live preview would be always on bleeding edge (because the bug appears not only with bob but also with dave & alice), that’s why when I found the same bug as in my local in there I decided to report it to you.

Sorry for the false alarm, Gene. I will be careful next time :P.

Thanks
